zproject/backends.py

```diff
@@ -1577,32 +1639,35 @@ def auth_complete(self, *args: Any, **kwargs: Any) -> Optional[HttpResponse]:
             and then change the RelayState param to the idp_name, because that's what
             SAMLAuth.auth_complete() expects.
         """
-        if 'RelayState' not in self.strategy.request_data():
-            logging.info("SAML authentication failed: missing RelayState.")
+        SAMLResponse = self.strategy.request_data().get('SAMLResponse')
+        if SAMLResponse is None:
+            logging.info("/complete/saml/: No SAMLResponse in request.")
             return None
 
-        # Set the relevant params that we transported in the RelayState:
-        redis_key = self.strategy.request_data()['RelayState']
-        relayed_params = self.get_data_from_redis(redis_key)
-        if relayed_params is None:
+        relayed_params = self.get_relayed_params()
+
+        subdomain = self.choose_subdomain(relayed_params)
+        if subdomain is None:
+            error_msg = "/complete/saml/: Can't figure out subdomain for this authentication request. " + \
+                "relayed_params: %s"
+            logging.info(error_msg, relayed_params)
             return None
 
-        idp_name = relayed_params.get("idp")
-        subdomain = relayed_params.get("subdomain")
-        if idp_name is None or subdomain is None:
-            error_msg = "Missing idp or subdomain value in relayed_params in SAML auth_complete: %s"
-            logging.info(error_msg, dict(subdomain=subdomain, idp=idp_name))
+        idp_name = self.get_issuing_idp(SAMLResponse)
+        if idp_name is None:
+            logging.info("/complete/saml/: No valid IdP as issuer of the SAMLResponse.")
             return None
 
         idp_valid = self.validate_idp_for_subdomain(idp_name, subdomain)
         if not idp_valid:
-            error_msg = "User authenticated with IdP %s but this provider is not " + \
-                        "enabled for this realm %s."
+            error_msg = "/complete/saml/: Authentication request with IdP %s but this provider is not " + \
+                        "enabled for this subdomain %s."
             logging.info(error_msg, idp_name, subdomain)
             return None
 
         result = None
         try:
+            self.strategy.session_set('subdomain', subdomain)
```

Comment:
Will this end up putting the value in the session twice?  I think it's still in `standard_relay_params`.

Reply:
I think subdomain generally wont be in the relayed params if it's different from the `subdomain` local variable, but that's indeed too fragile, so:
```
        try:
            self.strategy.session_set('subdomain', subdomain)
            for param, value in relayed_params.items():
                if param != 'subdomain' and param in self.standard_relay_params:
                    self.strategy.session_set(param, value)
```

Reply:
I think it's probably cleaner to just add the `subdomain` as a parameter to the list, and then run through the list:
```
@@ -1674,9 +1676,10 @@ class SAMLAuthBackend(SocialAuthMixin, SAMLAuth):
 
         result = None
         try:
-            self.strategy.session_set('subdomain', subdomain)
-            for param, value in relayed_params.items():
-                if param != 'subdomain' and param in self.standard_relay_params:
+            params = relayed_params.copy()
+            params['subdomain'] = subdomain
+            for param, value in params.items():
+                if param in self.standard_relay_params:
                     self.strategy.session_set(param, value)
 
             # super().auth_complete expects to have RelayState set to the idp_name,
```

So I'll merge that version.